collector/monitors/slashing_monitor.go

```diff
@@ -0,0 +1,125 @@
+)
+
+type SlashingMonitor struct {
+	metrics   map[Metric]float64
+	apiClient *client.TerraLiteForTerra
+	logger    *logrus.Logger
+}
+
+func NewSlashingMonitor(cfg config.CollectorConfig) *SlashingMonitor {
+	m := &SlashingMonitor{
+		metrics:   make(map[Metric]float64),
+		apiClient: cfg.GetTerraClient(),
+		logger:    cfg.Logger,
+	}
+
+	return m
+}
+
+func (m *SlashingMonitor) Name() string {
+	return "Slashing"
+}
+
+func (m *SlashingMonitor) InitMetrics() {
+	m.metrics = map[Metric]float64{
+		SlashingNumJailedValidators:     0,
+		SlashingNumTombstonedValidators: 0,
+		SlashingNumMissedBlocks:         0,
+	}
+}
+
+func (m *SlashingMonitor) Handler(ctx context.Context) error {
+	m.InitMetrics()
+
+	validatorAddresses, err := m.getValidatorPublicKeys()
+	if err != nil {
+		return fmt.Errorf("failed to get relevant validator addresses: %w", err)
+	}
+
+	for _, validatorAddress := range validatorAddresses {
+		signingInfoResponse, err := m.apiClient.Transactions.GetSlashingValidatorsValidatorPubKeySigningInfo(
+			&transactions.GetSlashingValidatorsValidatorPubKeySigningInfoParams{
+				ValidatorPubKey: validatorAddress,
+				Context:         ctx,
+			},
+		)
+		if err != nil {
+			return fmt.Errorf("failed to GetSlashingSigningInfos: %w", err)
+		}
+
+		var signingInfo = signingInfoResponse.GetPayload()
+
+		// This check is just in case, I haven't seen empty values in this
+		// field.
+		if len(signingInfo.JailedUntil) > 0 {
+			jailedUntil, err := time.Parse(jailTimeLayout, signingInfo.JailedUntil)
+			if err != nil {
+				m.logger.Errorf("failed to Parse `jailed_until` %s as %s: %s",
+					signingInfo.JailedUntil, jailTimeLayout, err)
+			} else {
+				// If the `jailed_until` property is set to a date in the future, this
+				// validator is jailed.
+				if jailedUntil.After(time.Now()) {
+					m.metrics[SlashingNumJailedValidators]++
+				}
+			}
+		}
+
+		// No blocks is sent as "", not as "0".
+		if len(signingInfo.MissedBlocksCounter) > 0 {
+			// If the current block is greater than minHeight and the validator's MissedBlocksCounter is
+			// greater than maxMissed, they will be slashed. So numMissedBlocks > 0 does not mean that we
+			// are already slashed, but is alarming. Note: Liveness slashes do NOT lead to a tombstoning.
+			// https://docs.terra.money/dev/spec-slashing.html#begin-block
+			numMissedBlocks, err := strconv.ParseInt(signingInfo.MissedBlocksCounter, 10, 64)
+			if err != nil {
+				m.logger.Errorf("failed to Parse `missed_blocks_counter:`: %s", err)
+			} else {
+				if numMissedBlocks > 0 {
+					m.metrics[SlashingNumMissedBlocks] += float64(numMissedBlocks)
+				}
+			}
+		}
+
+		if signingInfo.Tombstoned {
+			m.metrics[SlashingNumTombstonedValidators]++
+		}
+	}
+
+	return nil
+}
+
+func (m *SlashingMonitor) GetMetrics() map[Metric]float64 {
+	return m.metrics
+}
+
+func (m *SlashingMonitor) getValidatorPublicKeys() ([]string, error) {
+	// TODO: implement a query to the validators registry contract.
```

Comment:
Don't we need to do a query to the hub to get whitelisted validators (Query::WhitelistedValidators) before deployment of bassets v2? Btw, neither validators_registry nor the hub stores validators pubkeys. They store validators operator addresses only

Reply:
also true. added querying the whitelisted validators and then getting their consensus public keys